site/app/controllers/MiscController.php

```diff
@@ -58,4 +72,197 @@ private function display_file() {
             }
         }
     }
+
+    private function downloadFile() {
+        //Additional security
+        if (!($this->core->getUser()->accessGrading())) {
+            return false;
+        }
+        if (!file_exists($_REQUEST['path'])) {
+            throw new \InvalidArgumentException("File does not exist");
+        }
+        foreach (explode(DIRECTORY_SEPARATOR, $_REQUEST['path']) as $part) {
+            if ($part == ".." || $part == ".") {
+                throw new \InvalidArgumentException("Cannot have a part of the path just be dots");
+            }
+        }
+        $this->core->getOutput()->useHeader(false);
+        $this->core->getOutput()->useFooter(false);
+        $file_url = $_REQUEST['path'];
+        header('Content-Type: application/octet-stream');
+        header("Content-Transfer-Encoding: Binary"); 
+        header("Content-disposition: attachment; filename=\"" . basename($file_url) . "\""); 
+        readfile($file_url);
+    }
+
+    private function downloadZip() {
+        //Additional security
+        if (!($this->core->getUser()->accessGrading())) {
+            return false;
+        }
+        $zip_file_name = $_REQUEST['gradeable_id'] . "_" . $_REQUEST['user_id'] . "_" . date("m-d-Y") . ".zip";
+        $this->core->getOutput()->useHeader(false);
+        $this->core->getOutput()->useFooter(false);
+        $temp_dir = "/tmp";
+        //makes a random zip file name on the server
+        $temp_name = md5(uniqid($this->core->getUser()->getId(), true));
+        $zip_name = $temp_dir . "/" . $temp_name . ".zip";
+        $gradeable = $this->core->getQueries()->getGradeable($_REQUEST['gradeable_id'], $_REQUEST['user_id']);
+        $gradeable_path = $this->core->getConfig()->getCoursePath();
+        $active_version = $gradeable->getActiveVersion();
+        $folder_names = array();
+        $folder_names[] = "submissions";
+        $folder_names[] = "results";
+        $folder_names[] = "checkout";
+        $submissions_path = FileUtils::joinPaths($gradeable_path, $folder_names[0], $gradeable->getId(), $gradeable->getUser()->getId(), $active_version);
+        $results_path = FileUtils::joinPaths($gradeable_path, $folder_names[1], $gradeable->getId(), $gradeable->getUser()->getId(), $active_version);
+        $checkout_path = FileUtils::joinPaths($gradeable_path, $folder_names[2], $gradeable->getId(), $gradeable->getUser()->getId(), $active_version);
+        chdir ($temp_dir); //changes the directory to tmps
```

Comment:
Or, looking at `ZipArchive`, there's nothing stopping you from just giving an absolute path to `open()`. This avoids the use of `chdir()` altogether, eliminating a messy and potentially buggy instance of system-global state.